### PR TITLE
Filter out unknown catalog entities

### LINKED ISSUE #2813: Pin Catatlog Entity to Hotbar crash the app on restart
**Describe the bug**
Pin a Catalog Entity to Hotbar causes crashes

**To Reproduce**
Pin a Catalog Entity to Hotbar

Screen Shot 2021-05-19 at 2 52 49 PM

Confirm it's in Hotbar
Screen Shot 2021-05-19 at 2 54 49 PM

Restart Lens.

See this
Screen Shot 2021-05-19 at 2 57 40 PM

In console
Screen Shot 2021-05-19 at 2 58 41 PM

Above related to 
https://github.com/lensapp/lens/blob/master/src/renderer/components/hotbar/hotbar-menu.tsx#L55

If I only fix `hotbar-menu.tsx`

I will see this error in console instead

Screen Shot 2021-05-19 at 3 11 26 PM

This need fix too
https://github.com/lensapp/lens/blob/master/src/renderer/api/catalog-entity-registry.ts#L68

**Expected behavior**

No crashes

### PR DESCRIPTION
**Pull request by Nokel81**
- Keep raw data
- But filter unknown types until a category is registered
- Fixes a race condition.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

**Review by chenhunghan — DISMISSED** [review]
LGTM

**Review by chenhunghan — DISMISSED** [review]

**Review by jim-docker — APPROVED** [review]

File: src/common/utils/extended-map.ts [hunk 1]
```diff
@@ -22,32 +22,47 @@
...
     if (this.has(key)) {
-      return this.get(key);
+      throw new TypeError("Duplicate key in map");
     }
 
-    return this.set(key, getVal()).get(key);
+    return this.set(key, val);
   }
 
-  getOrDefault(key: K): V {
```
Comment (jakolehm) [inline src/common/utils/extended-map.ts:45]: 👍

File: src/renderer/api/__tests__/catalog-entity-registry.test.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ describe("CatalogEntityRegistry", () => {
         spec: {}
       }];
 
-      catalog.updateItems(items);
+      (catalog as any).rawItems.replace(items);
```
Comment (jakolehm) [inline src/renderer/api/__tests__/catalog-entity-registry.test.ts:45]: This feels bad, we could have testable subclass that allows replace.

File: src/renderer/api/catalog-entity-registry.ts [hunk 3]
```diff
@@ -49,23 +46,27 @@ export class CatalogEntityRegistry {
     return this._activeEntity;
   }
 
-  get items() {
-    return this._items;
+  @computed get items() {
```
Comment (jakolehm) [inline src/renderer/api/catalog-entity-registry.ts:49]: I think we should have tests that confirm filtering (now that it starts to be even more complex).

**Review by jakolehm — COMMENTED** [review]
This fixes problems on the receiving side (renderer) 👍 But why do we even sync "invalid" items from main to renderer?

**Review by jakolehm — APPROVED** [review]
I think we should merge this as is, I'll open follow-up PR for additional improvements.

File: src/common/catalog/catalog-category-registry.ts [hunk 4]
```diff
@@ -19,26 +19,38 @@
...
 
-  @action add(category: CatalogCategory) {
-    this.categories.push(category);
+  @action add(category: CatalogCategory): Disposer {
+    this.categories.add(category);
+
+    return () => this.categories.delete(category);
   }
 
-  @action remove(category: CatalogCategory) {
```
Comment (jakolehm) [inline src/common/catalog/catalog-category-registry.ts:32]: @Nokel81  was this removed by accident?

Comment (jakolehm) [reply src/common/catalog/catalog-category-registry.ts:32]: Maybe not, would have been good to mention that we are breaking the api here ... 🙂

**Comment by Nokel81** [comment]
> This fixes problems on the receiving side (renderer) 👍 But why do we even sync "invalid" items from main to renderer?

Because we sync everything from main to renderer and the category in renderer mighty not have been registered yet (race condition)

**Comment by jakolehm** [comment]
> > This fixes problems on the receiving side (renderer) 👍 But why do we even sync "invalid" items from main to renderer?
> 
> Because we sync everything from main to renderer and the category in renderer mighty not have been registered yet (race condition)

I get that but we need also protection on `main` side. Created a follow-up pr (#2824) that should improve this.

File: src/common/catalog/catalog-category-registry.ts (hunk 4 above)
Comment (Nokel81) [reply src/common/catalog/catalog-category-registry.ts:32]: Sorry, will start listing breaking API changes in my PR descriptions from now on.

**Pull request merged**